File: rb/features.rb
# LmWhatsapp SDK feature factory

require_relative 'feature/base_feature'
require_relative 'feature/debug_feature'
require_relative 'feature/idempotency_feature'
require_relative 'feature/metrics_feature'
require_relative 'feature/paging_feature'
require_relative 'feature/ratelimit_feature'
require_relative 'feature/retry_feature'
require_relative 'feature/test_feature'
require_relative 'feature/timeout_feature'


module LmWhatsappFeatures
  def self.make_feature(name)
    case name
    when "base"
      LmWhatsappBaseFeature.new
    when "debug"
      LmWhatsappDebugFeature.new
    when "idempotency"
      LmWhatsappIdempotencyFeature.new
    when "metrics"
      LmWhatsappMetricsFeature.new
    when "paging"
      LmWhatsappPagingFeature.new
    when "ratelimit"
      LmWhatsappRatelimitFeature.new
    when "retry"
      LmWhatsappRetryFeature.new
    when "test"
      LmWhatsappTestFeature.new
    when "timeout"
      LmWhatsappTimeoutFeature.new
    else
      LmWhatsappBaseFeature.new
    end
  end
end

File: py/lmwhatsapp_sdk/features.py
# LmWhatsapp SDK feature factory

from lmwhatsapp_sdk.feature.base_feature import LmWhatsappBaseFeature
from lmwhatsapp_sdk.feature.debug_feature import LmWhatsappDebugFeature
from lmwhatsapp_sdk.feature.idempotency_feature import LmWhatsappIdempotencyFeature
from lmwhatsapp_sdk.feature.metrics_feature import LmWhatsappMetricsFeature
from lmwhatsapp_sdk.feature.paging_feature import LmWhatsappPagingFeature
from lmwhatsapp_sdk.feature.ratelimit_feature import LmWhatsappRatelimitFeature
from lmwhatsapp_sdk.feature.retry_feature import LmWhatsappRetryFeature
from lmwhatsapp_sdk.feature.test_feature import LmWhatsappTestFeature
from lmwhatsapp_sdk.feature.timeout_feature import LmWhatsappTimeoutFeature


_FEATURES = {
    "base": lambda: LmWhatsappBaseFeature(),
    "debug": lambda: LmWhatsappDebugFeature(),
    "idempotency": lambda: LmWhatsappIdempotencyFeature(),
    "metrics": lambda: LmWhatsappMetricsFeature(),
    "paging": lambda: LmWhatsappPagingFeature(),
    "ratelimit": lambda: LmWhatsappRatelimitFeature(),
    "retry": lambda: LmWhatsappRetryFeature(),
    "test": lambda: LmWhatsappTestFeature(),
    "timeout": lambda: LmWhatsappTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
